x/capability/types/keys.go

```diff
@@ -0,0 +1,51 @@
+package types
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+const (
+	// ModuleName defines the module name
+	ModuleName = "capability"
+
+	// StoreKey defines the primary module store key
+	StoreKey = ModuleName
+
+	// MemStoreKey defines the in-memory store key
+	MemStoreKey = "mem_capability"
+)
+
+var (
+	// KeyIndex defines the key that stores the current globally unique capability
+	// index.
+	KeyIndex = []byte("index")
+
+	// KeyPrefixIndexCapability defines a key prefix that stores index to capability
+	// name mappings.
+	KeyPrefixIndexCapability = []byte("capability_index")
+)
+
+// RevCapabilityKey returns a reverse lookup key for a given module and capability
+// name.
+func RevCapabilityKey(module, name string) []byte {
+	return []byte(fmt.Sprintf("%s/rev/%s", module, name))
+}
+
+// FwdCapabilityKey returns a forward lookup key for a given module and capability
+// reference.
+func FwdCapabilityKey(module string, cap Capability) []byte {
+	return []byte(fmt.Sprintf("%s/fwd/%s", module, cap))
```

Comment:
Will this take the memory address of the capability? We don't want `.String()` to be called.

Reply:
(if `.String()` contains the memory reference it will work, but we don't need to put anything else in it)

Reply:
Yes, the `String` of a `Capability` should include the unique memory reference:

```go
func (ck *CapabilityKey) String() string {
	return fmt.Sprintf("CapabilityKey{%p, %d}", ck, ck.Index)
}
```

Now this is for a specific implementation (`CapabilityKey`). Whereas, `FwdCapabilityKey` takes an interface so we cannot guarantee any implementation adheres to this. That being said, to be 100% safe we can either:

1. Directly use the pointer in `FwdCapabilityKey`.
2. Remove the interface `Capability` and work directly with `CapabilityKey`. Throughout reading the spec and performing the implementation, it was not clear to me the need for an interface.

Reply:
I'm not sure we need an interface, actually, maybe we should try to remove it & simplify.

Reply:
I'll leave the interface for now. `FwdCapabilityKey` explicitly uses `%p` instead of `String`.